tests/parameter_flash_storage.cpp

```diff
@@ -0,0 +1,356 @@
+#include <CppUTest/TestHarness.h>
+#include <CppUTestExt/MockSupport.h>
+#include "../parameter_flash_storage.h"
+#include "../parameter_flash_storage_private.h"
+#include <parameter/parameter_msgpack.h>
+#include <cstdio>
+#include <cstring>
+#include "crc/crc32.h"
+
+TEST_GROUP(ConfigSaveTestCase)
+{
+    uint8_t data[128];
+    parameter_namespace_t ns;
+
+    void setup()
+    {
+        mock("flash").ignoreOtherCalls();
+        parameter_namespace_declare(&ns, NULL, NULL);
+    }
+};
+
+TEST(ConfigSaveTestCase, SavingConfigErasesPages)
+{
+    // In case we cannot find a valid block we erase the sector, just in case
+    mock("flash").expectOneCall("erase").withParameter("sector", data);
+
+    parameter_flash_storage_save(data, sizeof(data), &ns);
+}
+
+TEST(ConfigSaveTestCase, SavingConfigLockUnlock)
+{
+    mock("flash").strictOrder();
+
+    mock("flash").expectOneCall("unlock");
+    mock("flash").expectOneCall("erase").withParameter("sector", data);
+    mock("flash").expectOneCall("write"); // data
+    mock("flash").expectOneCall("write"); // CRC(len)
+    mock("flash").expectOneCall("write"); // len
+    mock("flash").expectOneCall("write"); // crc(data)
+    mock("flash").expectOneCall("lock");
+
+    parameter_flash_storage_save(data, sizeof(data), &ns);
+
+    // We need to manually check expectations since we tuned the flash mock to preserve ordering.
+    mock("flash").checkExpectations();
+}
+
+static void err_cb(void *p, const char *id, const char *err)
+{
+    (void) p;
+    (void) id;
+    (void) err;
+    char msg[128];
+    snprintf(msg, sizeof msg, "MessagePack error on item \"%s\": \"%s\"", id, err);
+
+    FAIL(msg);
+}
+
+TEST(ConfigSaveTestCase, SavingConfigWorks)
+{
+    // Declare a parameter and gives it a value.
+    parameter_t p;
+    parameter_integer_declare(&p, &ns, "foo");
+    parameter_integer_set(&p, 10);
+
+    // Check that the flash writer is used
+    // Number of expected written bytes is implementation-dependent
+    // Change if if necessary
+    for (int i = 0; i < 8; i++) {
+        mock("flash").expectOneCall("write");
+    }
```

Comment:
[suggestion] Can simplify to one line `mock("flash").expectNCalls(8, "write");`

Reply:
Woaw, did not know about that API, thanks!